// wdio-spec-reporter/build/reporter.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _Object$keys = require('babel-runtime/core-js/object/keys')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _humanizeDuration = require('humanize-duration');

var _humanizeDuration2 = _interopRequireDefault(_humanizeDuration);

var DURATION_OPTIONS = {
    units: ['m', 's'],
    round: true,
    spacer: ''
};

/**
 * Initialize a new `spec` test reporter.
 *
 * @param {Runner} runner
 * @api public
 */

var SpecReporter = (function (_events$EventEmitter) {
    _inherits(SpecReporter, _events$EventEmitter);

    function SpecReporter(baseReporter, config) {
        var options = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

        _classCallCheck(this, SpecReporter);

        _get(Object.getPrototypeOf(SpecReporter.prototype), 'constructor', this).call(this);

        this.baseReporter = baseReporter;
        this.config = config;
        this.options = options;
        this.shortEnglishHumanizer = _humanizeDuration2['default'].humanizer({
            language: 'shortEn',
            languages: { shortEn: {
                    h: function h() {
                        return 'h';
                    },
                    m: function m() {
                        return 'm';
                    },
                    s: function s() {
                        return 's';
                    },
                    ms: function ms() {
                        return 'ms';
                    }
                } }
        });

        this.errorCount = 0;
        this.indents = {};
        this.suiteIndents = {};
        this.specs = {};
        this.results = {};

        this.on('runner:start', function (runner) {
            this.suiteIndents[runner.cid] = {};
            this.indents[runner.cid] = 0;
            this.specs[runner.cid] = runner.specs;
            this.results[runner.cid] = {
                passing: 0,
                pending: 0,
                failing: 0
            };
        });

        this.on('suite:start', function (suite) {
            this.suiteIndents[suite.cid][suite.title] = ++this.indents[suite.cid];
        });

        this.on('test:pending', function (test) {
            this.results[test.cid].pending++;
        });

        this.on('test:pass', function (test) {
            this.results[test.cid].passing++;
        });

        this.on('test:fail', function (test) {
            this.results[test.cid].failing++;
        });

        this.on('suite:end', function (suite) {
            this.indents[suite.cid]--;
        });

        this.on('runner:end', function (runner) {
            this.printSuiteResult(runner);
        });
    }

    _createClass(SpecReporter, [{
        key: 'indent',
        value: function indent(cid, specUid) {
            var indents = this.suiteIndents[cid] && this.suiteIndents[cid][specUid] || -1;
            return indents < 0 ? '' : Array(indents).join('  ');
        }
    }, {
        key: 'getSymbol',
        value: function getSymbol(state) {
            var symbols = this.baseReporter.symbols;

            var symbol = '?'; // in case of an unknown state

            switch (state) {
                case 'pass':
                    symbol = symbols.ok;
                    break;
                case 'pending':
                    symbol = '-';
                    break;
                case 'fail':
                    this.errorCount++;
                    symbol = this.errorCount + ')';
                    break;
            }

            return symbol;
        }
    }, {
        key: 'getColor',
        value: function getColor(state) {
            var color = null; // in case of an unknown state

            switch (state) {
                case 'pass':
                case 'passing':
                    color = 'green';
                    break;
                case 'pending':
                    color = 'pending';
                    break;
                case 'fail':
                case 'failing':
                    color = 'fail';
                    break;
            }

            return color;
        }
    }, {
        key: 'getBrowserCombo',
        value: function getBrowserCombo(caps) {
            var verbose = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

            var device = caps.deviceName;
            var browser = caps.browserName;
            var version = caps.version || caps.platformVersion;
            var platform = caps.platform || caps.platformName;

            /**
             * mobile capabilities
             */
            if (device) {
                var program = (caps.app || '').replace('sauce-storage:', '') || caps.browserName;
                var executing = program ? 'executing ' + program : '';

                if (!verbose) {
                    return device + ' ' + platform + ' ' + version;
                }

                return (device + ' on ' + platform + ' ' + version + ' ' + executing).trim();
            }

            if (!verbose) {
                return (browser + ' ' + (version || '') + ' ' + (platform || '')).trim();
            }

            return browser + (version ? ' (v' + version + ')' : '') + (platform ? ' on ' + platform : '');
        }
    }, {
        key: 'getResultList',
        value: function getResultList(cid, suites) {
            var preface = arguments.length <= 2 || arguments[2] === undefined ? '' : arguments[2];

            var output = '';

            for (var specUid in suites) {
                var spec = suites[specUid];
                var indent = this.indent(cid, specUid);
                var specTitle = suites[specUid].title;
                output += preface + '   ' + indent + specTitle + '\n';

                for (var testUid in spec.tests) {
                    // Remove "before all" tests from the displayed results
                    if (specUid.indexOf('"before all"') === 0) {
                        continue;
                    }

                    var test = spec.tests[testUid];
                    var testTitle = spec.tests[testUid].title;

                    if (test.state === '') {
                        continue;
                    }

                    output += preface;
                    output += '       ' + indent;
                    output += this.baseReporter.color(this.getColor(test.state), this.getSymbol(test.state));
                    output += ' ' + testTitle + '\n';
                }

                output += preface.trim() + '\n';
            }

            return output;
        }
    }, {
        key: 'getSummary',
        value: function getSummary(states, duration) {
            var preface = arguments.length <= 2 || arguments[2] === undefined ? '' : arguments[2];

            var output = '';
            var displayedDuration = false;

            for (var state in states) {
                var testCount = states[state];
                var testDuration = '';

                /**
                 * don't display 0 passing/pending of failing test label
                 */
                if (testCount === 0) {
                    continue;
                }

                /**
                 * set duration
                 */
                if (!displayedDuration) {
                    testDuration = ' (' + this.shortEnglishHumanizer(duration, DURATION_OPTIONS) + ')';
                }

                output += preface + ' ';
                output += this.baseReporter.color(this.getColor(state), testCount);
                output += ' ' + this.baseReporter.color(this.getColor(state), state);
                output += testDuration;
                output += '\n';
                displayedDuration = true;
            }

            return output;
        }
    }, {
        key: 'getFailureList',
        value: function getFailureList(failures, preface) {
            var _this = this;

            var output = '';

            failures.forEach(function (test, i) {
                var title = typeof test.parent !== 'undefined' ? test.parent + ' ' + test.title : test.title;
                output += preface.trim() + '\n';
                output += preface + ' ' + _this.baseReporter.color('error title', i + 1 + ') ' + title + ':') + '\n';
                output += preface + ' ' + _this.baseReporter.color('error message', test.err.message) + '\n';
                if (test.err.stack) {
                    var stack = test.err.stack.split(/\n/g).map(function (l) {
                        return preface + ' ' + _this.baseReporter.color('error stack', l);
                    }).join('\n');
                    output += stack + '\n';
                } else {
                    output += preface + ' ' + _this.baseReporter.color('error stack', 'no stack available') + '\n';
                }
            });

            return output;
        }
    }, {
        key: 'getJobLink',
        value: function getJobLink(results, preface) {
            if (!results.config.host) {
                return '';
            }

            var output = '';
            if (results.config.host.indexOf('saucelabs.com') > -1) {
                output += preface.trim() + '\n';
                output += preface + ' Check out job at https://saucelabs.com/tests/' + results.sessionID + '\n';
                return output;
            }

            return output;
        }
    }, {
        key: 'getSuiteResult',
        value: function getSuiteResult(runner) {
            var cid = runner.cid;
            var stats = this.baseReporter.stats;
            var results = stats.runners[cid];
            var preface = '[' + this.getBrowserCombo(results.capabilities, false) + ' #' + cid + ']';
            var specHash = stats.getSpecHash(runner);
            var spec = results.specs[specHash];
            var combo = this.getBrowserCombo(results.capabilities);
            var failures = stats.getFailures().filter(function (f) {
                return f.cid === cid || _Object$keys(f.runner).indexOf(cid) > -1;
            });

            /**
             * don't print anything if no specs where executed
             */
            if (_Object$keys(spec.suites).length === 0) {
                return '';
            }

            this.errorCount = 0;
            var output = '';

            output += '------------------------------------------------------------------\n';
            output += preface + ' Session ID: ' + results.sessionID + '\n';
            output += preface + ' Spec: ' + this.specs[cid] + '\n';
            output += preface + ' Running: ' + combo + '\n';
            output += preface + '\n';
            output += this.getResultList(cid, spec.suites, preface);
            output += preface + '\n';
            output += this.getSummary(this.results[cid], spec._duration, preface);
            output += this.getFailureList(failures, preface);
            output += this.getJobLink(results, preface);
            output += preface + '\n';
            return output;
        }
    }, {
        key: 'printSuiteResult',
        value: function printSuiteResult(runner) {
            console.log(this.getSuiteResult(runner));
        }
    }]);

    return SpecReporter;
})(_events2['default'].EventEmitter);

exports['default'] = SpecReporter;
module.exports = exports['default'];
